Development.md

```diff
@@ -0,0 +1,273 @@
+
+Many API calls need to execute changes across subsystems. For example:
+
+* `convox build` needs to create a new Docker container for the build and collect its output and return code
+* `convox release promote` needs to perform a CloudFormation stack update
+
+Systems engineering best practices are encouraged:
+
+* Robust error handling
+* Logging that makes a developers life easier
+* Logging that can be turned into operational metrics (e.g. `count#push.retry=1)`
+* Code strategies that make it easy to simulate subsystem requests/responses in a test environment
+
+## Infrastructure Changes
+
+Racks apps and services are created, updated and destroyed via automated means. This is a DevOps best practice that minimizes human errors and accidents that cause downtime. On AWS this is accomplished with CloudFormation. Some examples of changes:
+
+* Rack should have a new option to provision and use private subnets
+* An app load balancer should have a new option to configure Proxy Protocol
+* `convox services create elasticsearch` should provision an ElasticSearch cluster
+
+Some general notes when making changes to the infrastructure templates:
+
+* Run `make -C api templates` to compile the templates and restart the webserver. The `templates.go` file updates should be checked in.
+* Run `make test` to exercise the app template regression tests. Changes to app.tmpl almost always need accompanying test changes.
+* Pay careful attention to both the update and rollback safety of changes. Rollbacks are extremely important for failure recovery.
+
+Convox uses [CloudFormation Custom Resources](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-custom-resources.html) by releasing `api/cmd/formation` as a Lambda handler that every Rack and App can use. This is very powerful, though often challenging to develop and debug.
+
+## Opening a Pull Request
+
+Rack follows the traditional [GitHub Flow](https://guides.github.com/introduction/flow/) where all changes start as a Pull Request.
+
+We encourage you to open a pull request for anything! For example:
+
+* A fully designed and tested enhancement
+* A untested but reasonable effort towards an enhancement
+* Work in progress (WIP) for questions or review
+
+The Rack maintainers aim to help land every reasonable pull request, and to provide clarity if a pull request can not be included.
+
+## Checklists
+
+Convox relies on checklists to safely and reliabily take code from a Pull Request to a published release. The standard release checklist is [https://github.com/convox/rack/blob/master/.github/PULL_REQUEST_TEMPLATE.md](PULL_REQUEST_TEMPLATE.md)
+
+We aim to automate and simplify the checklist over the life of the project to make releasing software easy, fast, and safe.
+
+## Release Changes for `convox rack update`
+
+The ultimate goal is to package changes so that a user can apply them with `convox rack update`. This involves:
+
+* Generate a release ID to tag every artifact with
```

Comment:
Inconsistent: All verbs should be gerunds:  Building Docker, Building and publishing Zip files, etc.